lib/kitchen/transport/winrm_file_transfer/remote_file.rb

```diff
@@ -0,0 +1,228 @@
+# -*- encoding: utf-8 -*-
+#
+# Author:: Matt Wrock (<matt@mattwrock.com>)
+#
+# Copyright (C) 2014, Matt Wrock
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "io/console"
+require "json"
+require "kitchen/transport/winrm_file_transfer/shell"
+
+module Kitchen
+  module Transport
+    module WinRMFileTransfer
+
+      # Represents a file to be copied to the test instance
+      # It maintains a winrm session to be used for all operations
+      #
+      # @author Matt Wrock <matt@mattwrock.com>
+      class RemoteFile
+
+        # @return [String] path to a file or directory to copy to the test instance
+        attr_reader :local_path
+
+        # @return [String] target path on test instance to copy to
+        attr_reader :remote_path
+
+        # @return [Boolean] true if the shell has been closed.
+        # The remote_file cannot be used if the shell is closed.
+        attr_reader :closed
+
+        # @return [WinRMWebService] the winrm connection used for upload
+        attr_reader :service
+
+        # @return [WinRMShell] the winrm shell used for upload
+        attr_reader :shell
+
+        # Initializes a new remote file intended for uploading
+        # to the test instance
+        #
+        # @param logger [Logger] kitchen logger for messaging output
+        # @param service [WinRMWebService] an active winrm connection
+        # @param local_path [String] path to a file or directory to
+        # copy to the test instance
+        # @param remote_path [String] target path on test instance to copy to
+        def initialize(logger, service, local_path, remote_path)
+          @closed = false
+          @service = service
+          @shell = Shell.new(logger, service)
+          @local_path = local_path
+          @remote_path = full_remote_path(local_path, remote_path)
+          @logger = logger
+          logger.debug("Creating RemoteFile of local '#{local_path}' at '#{@remote_path}'")
+        ensure
+          if !shell.nil?
+            ObjectSpace.define_finalizer(self, self.class.close(shell))
```

Comment:
This looked scary until I saw that this class had both close and self.close. Maybe rename to finalizer?